src/table.js

```diff
@@ -666,3 +667,90 @@ export function __table(source, operations) {
   }
   return source;
 }
+
+function isValidSchema(schema) {
+  if (!schema || !Array.isArray(schema)) return;
+  return schema.every((s) => {
+    s && typeof s.name === "string" && typeof s.type === "string";
+  });
+}
+
+function initKey() {
+  return {
+    other: 0,
+    boolean: 0,
+    integer: 0,
+    number: 0,
+    date: 0,
+    string: 0,
+    array: 0,
+    object: 0,
+    bigint: 0, // TODO for csv, tsv?
+    buffer: 0
+  };
+}
+
+function inferSchema(source) {
+  const schema = [];
+  const sampleSize = 100;
+  let sample = source.slice(0, sampleSize);
+  if (arrayIsPrimitive(sample)) {
+    sample = sample.map((d) => {
+      return {value: d};
+    });
+  }
+  const typeCounts = {};
+  sample.map((d) => {
+    for (const key in d) {
+      if (!typeCounts[key]) typeCounts[key] = initKey();
+      // for json and sqlite, we already have some types, but for csv and tsv, all
+      // columns are strings here.
+      const type = typeof d[key];
+      const value = type === "string" ? d[key]?.trim() : d[key];
+      if (value === null || value === undefined || value.length === 0)
+        typeCounts[key]["other"]++;
+      else if (type !== "string") {
+        if (Array.isArray(value)) typeCounts[key]["array"]++;
+        else if (value instanceof Date) typeCounts[key]["date"]++;
+        else if (value instanceof ArrayBuffer) typeCounts[key]["buffer"]++;
+        else if (type in typeCounts[key]) typeCounts[key][type]++; // number, bigint, boolean, or object
+      } else {
+        if (value === "true" || value === "false") typeCounts[key]["boolean"]++;
+        else if (!isNaN(+value) && /^-?[0-9]+$/.test(value))
```

Comment:
Number coercion happens implicitly when you call isNaN, so you can say `isNaN(value)` instead of `isNaN(+value)`. Would it be possible to only test number coercion once rather than doing it again below?